File: Python/LinkedLists/DoublyLinkedList/DoublyLinkedList.py
from Node import Node

class DoublyLinkedList():

    def __init__(self):
        self.head = None

    def push(self, data):
        if self.head:
            new_node = Node(data)
            new_node.nextnode = self.head
            self.head.prevnode = new_node
            self.head = new_node
        else:
            new_node = Node(data)
            self.head = new_node

    def append(self, data):
        if self.head:
            temp = self.head
            while temp.nextnode:
                temp = temp.nextnode
            new_node = Node(data)
            temp.nextnode = new_node
            new_node.prevnode = temp
        else:
            new_node = Node(data)
            self.head = new_node

    def insert_after(self, prev_data, data):
        if self.head:
            temp = self.head
            while temp and temp.getNodeValue() != prev_data:
                temp = temp.nextnode
            if temp:
                new_node = Node(data)
                new_node.nextnode = temp.nextnode
                temp.nextnode = new_node
                new_node.prevnode = temp
                if new_node.nextnode:
                    new_node.nextnode.prevnode = new_node
            else:
                print("Listede Boyle Bir Eleman Yok!")
        else:
            print("Listede Boyle Bir Eleman Yok!")

    def delete(self,data):
        if self.head:
            temp = self.head
            while temp and temp.getNodeValue() != data:
                temp = temp.nextnode
            if temp:
                if temp == self.head:
                    if temp.nextnode:
                        temp.nextnode.prevnode = None
                        self.head = temp.nextnode
                        temp = None
                    else:
                        self.head = None
                else:
                    temp.prevnode.nextnode = temp.nextnode
                    if temp.nextnode:
                        temp.nextnode.prevnode = temp.prevnode
                    temp = None
            else:
                print("Listede Boyle Bir Eleman Yok!")
        else:
            print("Listede Eleman Yok!")

    def print_list(self):
        if self.head:
            temp = self.head
            while temp:
                print(temp.data)
                temp = temp.nextnode
        else:
            print("Listede Eleman Yok!")

linked_list = DoublyLinkedList()
linked_list.push(14)
linked_list.push(3)
linked_list.push(5)
linked_list.push(17)
linked_list.print_list()
linked_list.append(12)
linked_list.append(77)
linked_list.append(91)
linked_list.print_list()
linked_list.insert_after(77,15)
linked_list.insert_after(3,13)
linked_list.insert_after(91,68)
linked_list.insert_after(68,69)
linked_list.insert_after(17,1)
linked_list.insert_after(19,12)
linked_list.print_list()
print("--------------------")
linked_list.delete(999)
linked_list.delete(77)
linked_list.delete(68)
linked_list.delete(5)
linked_list.delete(69)
linked_list.delete(91)
linked_list.delete(17)
linked_list.append(91)

linked_list.print_list()

File: C++/LinkedLists/DoublyLinkedList/DoublyLinkedList.cpp
#include <iostream>

using namespace std;

struct node{
    int n;
    node* next;
    node* prev;
};

void append(node* root, int x){
    node* new_node = new node[1];
    new_node->n = x;
    node* iter = root;
    while(iter->next!=nullptr){
        iter=iter->next;
    }
    iter->next = new_node;
    new_node->prev = iter;
    new_node->next = nullptr;
}

node* add_front(node* root, int x){
    node* new_node = new node[1];
    new_node->n = x;
    new_node->prev=nullptr;
    new_node->next=root;
    root->prev=new_node;
    root = new_node;
    return root;
}

node* remove(node* root,int x){
    node* iter = root;
    while(iter != nullptr && iter->n != x){
        iter=iter->next;
    }
    if(iter == nullptr){
        cout<<"Listede Boyle Bir Eleman Yok!";
    }
    else{
    	if(iter->prev == nullptr){
    		root = iter->next;
		}
		else if(iter->next == nullptr){
			iter->prev->next = nullptr;
		}
		else{
			iter->prev->next = iter->next;
        	iter->next->prev = iter->prev;
		}
        delete[] iter;
        return root;
    }
}

void insert_after(node* root, int x, int y){
    node* iter = root;
    while(iter != nullptr && iter->n != x){
        iter = iter->next;
    }
    if(iter == nullptr){
        cout<<"Listede Boyle Bir Eleman Yok!";
    }
    else{
        node* new_node = new node[1];
        new_node->n = y;
        new_node->next = iter->next;
        if(iter->next != nullptr)
			iter->next->prev = new_node;
        iter->next = new_node;
        new_node->prev = iter;
    }
}

void print_list(node* root){
	cout<<endl<<"-------"<<endl;
    node* iter = root;
    while(iter!=nullptr){
        cout<<endl<<iter->n<<endl;
        iter=iter->next;
    }
    cout<<endl<<"-------"<<endl;
}

int main(){
    node* root;
    root = new node[1];
    root->n = 0;
    root->next = nullptr;
    root->prev = nullptr;
//    for(int i = 1; i <= 5; i++){
//        append(root,i);
//    }
	for(int i=1;i<=5;i++){
		root=add_front(root,i);
	}
	append(root,8);
	root = remove(root,8);
    append(root,8);
    root = add_front(root,10);
    append(root,13);
    root = remove(root,10);
    root = remove(root,3);
	print_list(root);
    root = add_front(root,15);
    root = add_front(root,19);
    root = add_front(root,38);
    print_list(root);
    remove(root,3);
    remove(root,19);
    remove(root,28);
    print_list(root);
    cout<<"-------"<<endl;
    insert_after(root, 4, 7);
    insert_after(root, 7, 88);
    insert_after(root, 38, 93);
	insert_after(root, 15, 103);
	insert_after(root, 13, 104);
	root = remove(root,104);
	append(root,16);
	root = add_front(root,158);
	insert_after(root,158,77);
    print_list(root);

    return 0;
}
